libcxx/include/map

```diff
@@ -972,31 +972,15 @@ public:
 
   _LIBCPP_HIDE_FROM_ABI map(const map& __m) : __tree_(__m.__tree_) { insert(__m.begin(), __m.end()); }
 
-  _LIBCPP_HIDE_FROM_ABI map& operator=(const map& __m) {
-#  ifndef _LIBCPP_CXX03_LANG
-    __tree_ = __m.__tree_;
-#  else
-    if (this != std::addressof(__m)) {
```

Comment:
It's not clear to me that before/after are equivalent in non-C++03 mode -- can you explain why they are?

Also, I had played around this area in https://reviews.llvm.org/D121485 (which never landed), can you take a look and ensure that concerns back then don't apply to this? Finally, can you benchmark before/after? In principle, we shouldn't see a difference IIUC?